contracts/token/ERC721/ERC721Token.sol

```diff
@@ -209,6 +152,46 @@ contract ERC721Token is ERC721 {
     ownedTokens[_from].length--;
     ownedTokensIndex[_tokenId] = 0;
     ownedTokensIndex[lastToken] = tokenIndex;
-    totalTokens = totalTokens.sub(1);
   }
+
+  /**
+  * @dev Internal function to mint a new token
+  * @dev Reverts if the given token ID already exists
+  * @param _to address the beneficiary that will own the minted token
+  * @param _tokenId uint256 ID of the token to be minted by the msg.sender
+  */
+  function doMint(address _to, uint256 _tokenId) internal {
+    super.doMint(_to, _tokenId);
+    
+    allTokensIndex[_tokenId] = allTokens.length;
+    allTokens.push(_tokenId);
+  }
+
+  /**
+  * @dev Internal function to burn a specific token
+  * @dev Reverts if the token does not exist
+  * @param _owner owner of the token to burn
+  * @param _tokenId uint256 ID of the token being burned by the msg.sender
+  */
+  function doBurn(address _owner, uint256 _tokenId) internal {
+    super.doBurn(_owner, _tokenId);
+
+    // Clear metadata (if any)
+    if (bytes(tokenURIs[_tokenId]).length == 0) {
```

Comment:
`==` should be `!=`, but I'd even leave the check out. We can just `delete`, it might actually be cheaper.

I don't understand the cast to `bytes`.

Reply:
Strings don't have a "length" property, which is why it is cast to bytes.

Reply:
Deleting without checking (if there was no value set) costs 10x more gas than checking beforehand. Tested with this contract in Remix:

```solidity
pragma solidity "0.4.21";

contract Foo {
    mapping(uint256 => string) internal tokenURIs;
    
    function set(uint id, string value) public {
        tokenURIs[id] = value;
    }
    
    function deleteWithoutCheck(uint id) public {
        delete tokenURIs[id];
    }
    
    function deleteWithCheck(uint id) public {
        if (bytes(tokenURIs[id]).length != 0) {
            delete tokenURIs[id];
        }
    }
}
```

Reply:
And thanks for catching the `!=` vs `==`!